libq/cur_sys_pkg.c

```diff
@@ -0,0 +1,348 @@
+#include "config.h"
+
+#include <openssl/md5.h>
+#include <openssl/sha.h>
+#include <assert.h>
+#include <sys/stat.h> 
+#include <sys/types.h>
+#include <dirent.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdio.h>
+#include <xalloc.h>
+
+#include "atom.h"
+#include "hash.h"
+#include "xchdir.h"
+#include "cur_sys_pkg.h"
+
+
+#define HASH_SIZE 32
+#define SIZE_STR_VAR_DB_PKG 12
+
+//private
+
+//data
+typedef struct cur_pkg_tree_node {
+  char *key;
+  char *hash_buffer;
+  char *package_name;
+  unsigned int safe_to_free_package_name;
+  struct cur_pkg_tree_node *greater;
+  struct cur_pkg_tree_node *minor;
+}cur_pkg_tree_node;
+
+
+//functions
+static unsigned int conv_char_int(char dig)
+{
+  if((int) dig > 57) 
+  {
+    return (int)dig - 87;
+  }
+  return (int )dig - 48;
+}
+
+static int compare_hash_num(char *hash1,char*hash2)
+{
+  int temp1,temp2;
+  for(int i=0;i<HASH_SIZE;++i)
+  {
+    temp1=conv_char_int(hash1[i]);
+    temp2=conv_char_int(hash2[i]);
+    if(temp2 > temp1)
+    {
+      return 1;
+    }else if (temp2 < temp1) {
+      return -1;
+    }
+  }
+  return 0;
+}
+
+static char *get_fullname_package(depend_atom *datom)
+{
+  int cat_len,name_len=0;
+  char *package_name =NULL;
+
+  assert(datom!=NULL);
+  assert(datom->CATEGORY!=NULL);
+  assert(datom->PN!=NULL);
+
+  cat_len=strlen(datom->CATEGORY);
+  name_len=strlen(datom->PN);
+  package_name=calloc((cat_len +1+ name_len +1), sizeof(*package_name));
+  package_name[cat_len + 1 +name_len]='\0';
+
+  strcat(package_name,datom->CATEGORY);
+  strcat(package_name,"/");
+  strcat(package_name,datom->PN);
```

Comment:
Hmmm, `atom_format("%{CAT}/%{PN}", datom)` isn't sufficient here?

Reply:
i didn't see that, thx. I will check